### PR TITLE
Get and use IBCMerge tooling from internal/authed VSTS feed

### PR DESCRIPTION
**Pull request by dagood**
This adds "internal-tool-runtime" restore just before Sync for tooling that needs to come from an authed NuGet feed. On Windows with full framework, a NuGet.Config is created on disk with an encrypted password. On *nix with Core, the unimplemented approach is to pipe a in-memory NuGet.Config into the restore command that has the credentials in plaintext. (The type of user encryption used on Windows isn't supported xplat, and the piping workaround doesn't work on Windows.) The hope in the future is to implement both methods as part of the bootstrap flow, once CoreFX uses it.

IBCMerge is the first tool that we need to restore this way. I changed the optimization targets to pull it in using the new internal tool restore flow. I made some changes to support the new file type we're using to move IBC data into the build. The IBC data package id, version, and source are all configurable by the repository using the optimization targets, so auto-upgrade can happen independent of buildtools.

---

I tested this out by cloning the Windows leg build definition in CoreFX, adding the needed secrets/properties, and running it. It resulted in signed DLLs with IBC data merged, where it exists. (See "Packages" artifact here: https://devdiv.visualstudio.com/DevDiv/_build/index?buildId=558294&_a=summary&tab=artifacts.)

I have a `Microsoft.DotNet.Internal.IBCMerge` `1.0.0-alpha-00001` package in a VSTS private feed "dotnet-core-internal-tooling-test" right now, and it has `lib/net45/ibcmerge.exe`. This makes it easy to pick up with `PrereleaseResolveNuGetPackageAssets`, but I'm open to change that if it seems wrong.

@chcosta @MattGal @AlexGhiondea PTAL

@weshaggard FYI I'm planning on using this to start merging IBC data in the CoreFX master official build.

### CODE REVIEW

**Comment by dagood** [comment]
Fixed merge conflict, updated IBCMerge package to the `4.6.0-alpha-00001` one I published to the "real" internal feed that I plan to use. PTAL

**Comment by dagood** [comment]
@dotnet-bot test this please

File: src/Microsoft.DotNet.Build.Tasks/GenerateEncryptedNuGetConfig.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            List<PackageSource> sourceList = sourceProvider.LoadPackageSources().ToList();
+
+            foreach (ITaskItem newSourceItem in Sources)
+            {
+                string name = newSourceItem.ItemSpec;
+                string uri = newSourceItem.GetMetadata("Uri");
+                if (string.IsNullOrEmpty(uri))
+                {
+                    throw new ArgumentException(
+                        $"Item '{newSourceItem}': 'Uri' metadata required, found '{uri}'.");
```
Comment (MattGal) [inline src/Microsoft.DotNet.Build.Tasks/GenerateEncryptedNuGetConfig.cs]: If it's null or empty, not much point in printing it.

File: src/Microsoft.DotNet.Build.Tasks/PackageFiles/codeOptimization.targets [hunk 2]
```diff
@@ -3,39 +3,34 @@
...
-          Condition="'$(OSEnvironment)'=='Windows_NT' and '$(EnableProfileGuidedOptimization)'=='true' and Exists('$(OptimizationDataDir)/$(AssemblyName).zip')">
+  <Target Name="OptimizeWithTrainingData"
+          AfterTargets="AfterBuild"
+          BeforeTargets="CopyFilesToOutputDirectory"
+          DependsOnTargets="RestoreOptimizationDataPackage;ResolveInternalTools"
+          Condition="'$(OS)'=='Windows_NT' and '$(EnableProfileGuidedOptimization)'=='true' and Exists('$(OptimizationDataDir)$(AssemblyName).pgo')">
+
+    <!-- Find IBCMerge as a resolved internal tool. -->
+    <PropertyGroup>
+      <IBCMergeToolPath Condition="'%(Filename)%(Extension)'=='ibcmerge.exe'">@(ResolvedInternalToolReferences)</IBCMergeToolPath>
```
Comment (MattGal) [inline src/Microsoft.DotNet.Build.Tasks/PackageFiles/codeOptimization.targets]: Not feedback, but this is a cool way to find a file path in msbuild!

File: src/Microsoft.DotNet.Build.Tasks/GenerateEncryptedNuGetConfig.cs (hunk 1 above)
Comment (dagood) [inline src/Microsoft.DotNet.Build.Tasks/GenerateEncryptedNuGetConfig.cs]: Good point, removed.

**Pull request merged**